### ISSUE TITLE
Rclone size reports incorrect size if --fast-list is used ( default behaviour )

### ISSUE THREAD

**Issue by xd003** [rclone/rclone#4289]
#### What is the problem you are having with rclone?

```rclone size remote:``` reports incorrect size for non team drive remotes if --fast-list is used . If i use ```rclone size remote: --disable ListR``` , i am getting accurate results

#### What is your rclone version (output from `rclone version`)

rclone v1.52.0

#### Which OS you are using and how many bits (eg Windows 7, 64 bit)

- os/arch: windows/386
- go version: go1.14.3

####  Which cloud storage system are you using? (eg Google Drive)

Google Drive

#### A log from the command with the `-vv` flag (eg output from `rclone -vv copy /tmp remote:tmp`)
**Actual size - 2.853 gb**
**Test with rclone size remote:**
```
PS C:\rclone> rclone size test: -vv
2020/05/28 22:57:48 DEBUG : rclone: Version "v1.52.0" starting with parameters ["C:\\rclone\\rclone.exe" "size" "test:" "-vv"]
2020/05/28 22:57:48 DEBUG : Using config file from "C:\\Users\\Sanjeev\\.config\\rclone\\rclone.conf"
Total objects: 63
Total size: 292.184 MBytes (306377599 Bytes)
2020/05/28 22:57:52 DEBUG : 18 go routines active
PS C:\rclone> rclone size test: -vv
2020/05/28 22:58:01 DEBUG : rclone: Version "v1.52.0" starting with parameters ["C:\\rclone\\rclone.exe" "size" "test:" "-vv"]
2020/05/28 22:58:01 DEBUG : Using config file from "C:\\Users\\Sanjeev\\.config\\rclone\\rclone.conf"
Total objects: 74
Total size: 334.623 MBytes (350877842 Bytes)
2020/05/28 22:58:03 DEBUG : 18 go routines active
```
As you can see its incorrect and even reports inconsistent size randomly everytime

**Test with rclone size remote: --disable ListR**
```
PS C:\rclone> rclone size test: --disable ListR -vv
2020/05/28 23:05:18 DEBUG : rclone: Version "v1.52.0" starting with parameters ["C:\\rclone\\rclone.exe" "size" "test:" "--disable" "ListR" "-vv"]
2020/05/28 23:05:18 DEBUG : Using config file from "C:\\Users\\Sanjeev\\.config\\rclone\\rclone.conf"
2020/05/28 23:05:18 DEBUG : Reset feature "ListR"
Total objects: 354
Total size: 2.853 GBytes (3063873067 Bytes)
2020/05/28 23:05:31 DEBUG : 18 go routines active
PS C:\rclone> rclone size test: --disable ListR -vv
2020/05/28 23:05:38 DEBUG : rclone: Version "v1.52.0" starting with parameters ["C:\\rclone\\rclone.exe" "size" "test:" "--disable" "ListR" "-vv"]
2020/05/28 23:05:38 DEBUG : Using config file from "C:\\Users\\Sanjeev\\.config\\rclone\\rclone.conf"
2020/05/28 23:05:38 DEBUG : Reset feature "ListR"
Total objects: 354
Total size: 2.853 GBytes (3063873067 Bytes)
2020/05/28 23:05:40 DEBUG : 18 go routines active
```
As you can see --disabling fast list gives correct results
Although please note that fast list doesnt seem bugged for team drive remotes and personal drives , so anyone who wanna reproduce this , dont test it on a team drive or personal drive remote
 It could be someone else's Gdrive folder which is public or shared with you
For someone who wanna quickly test it , i am dropping my rclone config file , just replace your token
```
[test]
type = drive
scope = drive
root_folder_id = 1vbyJ26ojnw5NvEpL2mrRkl5N--zqc1yZ
token = {"access_token":#####"}
```
( the google folder is public so anyone can test it themselves  its the very same 2.853 gb sized folder used in my test remote  )

**NOTE** - _The issue is not with rclone size but rather with --fast-list flag . Even if i do a rclone copy with --fast-list flag enabled for the above remote , it skips files just like it does while rclone size .I have chosen rclone size to show bugged --fast-list because its quicker than reproducing with rclone copy

**Comment by ncw** [comment]
This is potentially related to this issue: https://github.com/rclone/rclone/issues/3114

Do you know if the drive is a shared/team drive?

**Comment by xd003** [comment]
> This is potentially related to this issue: #3114
> 
> Do you know if the drive is a shared/team drive?

since the folder is publically available , it definitely isnt a team drive , its just a public folder from someone else's google account . This issue can also be reproduced in shared folders ( not td )

**Comment by ncw** [comment]
This looks like the same issue as #3114

Do do ListR rclone does queries of the form 

```
2020/05/29 17:19:14 DEBUG : list query = trashed=false and
('1nJBJzllQF63jA2kMPwkwtCTUV7PD_tkR' in parents or '1yIWAJcjkUOhppRy2Tppw1wFO-nqEDFX7' in parents)
```

However these just return nothing

```
2020/05/29 17:19:14 DEBUG : HTTP REQUEST (req 0xc00028c500)
2020/05/29 17:19:14 DEBUG : GET /drive/v3/files?alt=json&fields=files%28id%2Cname%2Csize%2Cmd5Checksum%2Ctrashed%2CmodifiedTime%2CcreatedTime%2CmimeType%2Cparents%2CwebViewLink%2CshortcutDetails%29%2CnextPageToken%2CincompleteSearch&includeItemsFromAllDrives=true&pageSize=1000&prettyPrint=false&q=trashed%3Dfalse+and+%28%271nJBJzllQF63jA2kMPwkwtCTUV7PD_tkR%27+in+parents+or+%271yIWAJcjkUOhppRy2Tppw1wFO-nqEDFX7%27+in+parents%29&supportsAllDrives=true HTTP/1.1
Host: www.googleapis.com
User-Agent: rclone/v1.52.0-DEV
Authorization: XXXX
X-Goog-Api-Client: gl-go/1.14.0 gdcl/20200410
Accept-Encoding: gzip

2020/05/29 17:19:14 DEBUG : >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
2020/05/29 17:19:14 DEBUG : <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
2020/05/29 17:19:14 DEBUG : HTTP RESPONSE (req 0xc00028c500)
2020/05/29 17:19:14 DEBUG : HTTP/1.1 200 OK
Transfer-Encoding: chunked
Alt-Svc: h3-27=":443"; ma=2592000,h3-25=":443"; ma=2592000,h3-T050=":443"; ma=2592000,h3-Q050=":443"; ma=2592000,h3-Q049=":443"; ma=2592000,h3-Q048=":443"; ma=2592000,h3-Q046=":443"; ma=2592000,h3-Q043=":443"; ma=2592000,quic=":443"; ma=2592000; v="46,43"
Cache-Control: private, max-age=0, must-revalidate, no-transform
Content-Security-Policy: frame-ancestors 'self'
Content-Type: application/json; charset=UTF-8
Date: Fri, 29 May 2020 16:19:14 GMT
Expires: Fri, 29 May 2020 16:19:14 GMT
Server: GSE
Vary: Origin
Vary: X-Origin
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
X-Xss-Protection: 1; mode=block

25
{"incompleteSearch":false,"files":[]}
0
```

These queries work fine normally if done one ID at a time (eg when you `--disable ListR`).

I tried these queries in the [API explorer](https://developers.google.com/drive/api/v3/reference/files/list?apix_params=%7B%22includeItemsFromAllDrives%22%3Atrue%2C%22includeTeamDriveItems%22%3Atrue%2C%22q%22%3A%22((%271nJBJzllQF63jA2kMPwkwtCTUV7PD_tkR%27%20in%20parents)%20or%20(%271yIWAJcjkUOhppRy2Tppw1wFO-nqEDFX7%27%20in%20parents))%22%2C%22supportsAllDrives%22%3Atrue%2C%22supportsTeamDrives%22%3Afalse%7D) with the same result..

I'm thinking that this is a bug in Google Drive unless we can think of some query parameter I've missed

It looks like a bug has been reported about this or something very similar

https://issuetracker.google.com/issues/149522397

I think rclone needs to engage a work-around for this....

**Comment by ncw** [comment]
I had a go at making a workaround here - can you have a test?

https://beta.rclone.org/branch/v1.52.0-004-g96c6137c-fix-4289-listr-shared-beta/ (uploaded in 15-30 mins)

    drive: fix missing items when listing using --fast-list / ListR
    
    This is caused by a bug in Google drive where, in some circumstances
    querying for "(A in parents) or (B in parents)" returns nothing
    whereas querying for "A in parents" and "B in parents" separately
    works fine.
    
    This has been reported here:
    
    https://issuetracker.google.com/issues/149522397
    
    This workaround detects this condition by seeing if a listing for more
    than one directory at once returns nothing.
    
    If it does then it retries each one individually.
    
    This can potentially have a false positive if the user has multiple
    empty directories which are queried at once. The consequence of this
    will be that ListR is disabled for a while until the directories are
    found to be actually empty in which case it will be re-enabled.

**Comment by xd003** [comment]
> I had a go at making a workaround here - can you have a test?
> 
> https://beta.rclone.org/branch/v1.52.0-004-g96c6137c-fix-4289-listr-shared-beta/ (uploaded in 15-30 mins)
> 
> ```
> drive: fix missing items when listing using --fast-list / ListR
> 
> This is caused by a bug in Google drive where, in some circumstances
> querying for "(A in parents) or (B in parents)" returns nothing
> whereas querying for "A in parents" and "B in parents" separately
> works fine.
> 
> This has been reported here:
> 
> https://issuetracker.google.com/issues/149522397
> 
> This workaround detects this condition by seeing if a listing for more
> than one directory at once returns nothing.
> 
> If it does then it retries each one individually.
> 
> This can potentially have a false positive if the user has multiple
> empty directories which are queried at once. The consequence of this
> will be that ListR is disabled for a while until the directories are
> found to be actually empty in which case it will be re-enabled.
> ```

Just tested in this beta build
```
PS C:\rclone> rclone size test: -vv
2020/05/31 09:04:53 DEBUG : rclone: Version "v1.52.0-004-g96c6137c-fix-4289-listr-shared-beta" starting with parameters ["C:\\rclone\\rclone.exe" "size" "test:" "-vv"]
2020/05/31 09:04:53 DEBUG : Using config file from "C:\\Users\\Sanjeev\\.config\\rclone\\rclone.conf"
2020/05/31 09:04:55 NOTICE: Google drive root '': Disabling ListR to work around bug in drive as multi listing (2) returned no entries
2020/05/31 09:04:55 DEBUG : Google drive root '': Recycled 2 entries
2020/05/31 09:04:55 DEBUG : Google drive root '': Recycled 13 entries
Total objects: 354
Total size: 2.853 GBytes (3063873067 Bytes)
2020/05/31 09:04:59 DEBUG : 18 go routines active
```
i see , now it automatically disables ListR to avoid giving incorrect results

**Comment by ncw** [comment]
> Yes its fine now, it automatically disables ListR to avoid giving incorrect results

Great!

> is this fastlist disabling check done globally for other rclone commands that accept fast list flag too ? because if lets say someone is doing rclone copy from remote where fastlist is bugged and he explicitly uses fast list flag , it will skip files while cloning

Yes it will. `ListR` is the internal command that `--fast-list` uses so it will cover all cases. Some commands like `rclone size` use `ListR` without the `--fast-list` flag because it doesn't use any more memory.

I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org) in 15-30 mins and released in v1.52.1
